contract/models/contract.py

```diff
@@ -213,13 +213,73 @@ def action_contract_send(self):
+        not possible to create an invoice line with no account
+
+        :param invoice_values: dictionary (invoice values)
+        :return: updated dictionary (invoice values)
+        """
+        # If no account on the product, the invoice lines account is
+        # taken from the invoice's journal in _onchange_product_id
+        # This code is not in finalize_creation_from_contract because it's
+        # not possible to create an invoice line with no account
+        new_invoice = self.env['account.invoice'].new(invoice_values)
+        for invoice_line in new_invoice.invoice_line_ids:
+            name = invoice_line.name
+            account_analytic_id = invoice_line.account_analytic_id
+            price_unit = invoice_line.price_unit
+            invoice_line.invoice_id = new_invoice
+            invoice_line._onchange_product_id()
+            invoice_line.update({
+                'name': name,
+                'account_analytic_id': account_analytic_id,
+                'price_unit': price_unit,
+            })
+        return new_invoice._convert_to_write(new_invoice._cache)
+
+    @api.model
+    def _finalize_invoice_creation(self, invoices):
+        for invoice in invoices:
+            invoice._onchange_partner_id()
+        invoices.compute_taxes()
+
+    @api.model
+    def _finalize_and_create_invoices(self, invoices_values):
+        """
+        This method:
+         - finalizes the invoices values (onchange's...)
+         - creates the invoices
+         - finalizes the created invoices (onchange's, tax computation...)
+        :param invoices_values: list of dictionaries (invoices values)
+        :return: created invoices (account.invoice)
+        """
+        if isinstance(invoices_values, dict):
+            invoices_values = [invoices_values]
+        final_invoices_values = []
+        for invoice_values in invoices_values:
+            final_invoices_values.append(
+                self._finalize_invoice_values(invoice_values))
+        invoices = self.env['account.invoice'].create(final_invoices_values)
+        self._finalize_invoice_creation(invoices)
+        return invoices
+
+    @api.model
+    def _get_contracts_to_invoice_domain(self, date_ref=None):
+        """
+        This method builds the domain to use to find all
+        contracts (account.analytic.account) to invoice.
+        :param date_ref: optional reference date to use instead of today
+        :return: list (domain) usable on account.analytic.account
+        """
         domain = []
-        date_ref = fields.Date.context_today(self)
-        if contract:
-            contract.ensure_one()
-            date_ref = contract.recurring_next_date
-            domain.append(('id', '=', contract.id))
+        if not date_ref:
+            date_ref = fields.Date.context_today(self)
         domain.extend(
             [
                 ('recurring_invoices', '=', True),
```

Comment:
@ThomasBinsfeld add recurring_next_date != False

@sbejaoui would this be sufficient to filter out contracts that have nothing to invoice anymore because they are finished, canceled or whatever else?

Reply:
@sbidoul , I think it's sufficient for contracts. I see that we filter contract line with is_cancel, so it's good enough for me.

Reply:
@sbejaoui and when we check is_cancel on a line, is it correct that recurring_next_date is unset on the line, and when it is unset on all lines it is also unset on the contract?

Reply:
Sure, when recurring_next_date is unset on all lines it's unset on the contract.

For now, cancel() does not unset recurring_next_date, but uncancel require a new one. So thinking about it, we can unset it on cancel